streaming-payments.md

```diff
@@ -0,0 +1,375 @@
+xPayHeader = [ amount, shopAddress + '.' + userId, base64(conditionSeed) ]
+paymentId = '1'
+paymentPacketContents = {
+  amount: xPayHeader[0],
+  account: xPayHeader[1] + '.' + paymentId,
+  data: ''
+}
+fulfillment = hmac(conditionSeed, paymentPacketContents.account)
+condition = sha256(fulfillment)
+```
+
+To run this new version of the shop, type this into your terminal:
+
+```sh
+npm install ilp-packet
+node ./shop2.js
+```
+
+The ILP paacket is not very useful in a one-transfer payment, but in future tutorials, we see how connectors can forward an
+incoming transfer, and thus connect one ledger to other ledgers. When that happens, we say the whole sender
+to receiver process is the "payment", and each link in the chain is a "transfer", so one payment consists of 
+one or more transfers. The sender can then put information in the ILP packet which the receiver can use
+to generate the fulfillment, and if that information would be tempered with by connectors along the way,
+this would be detected because the fulfillment would not match.
+The ILP Payment Packet is serialized into OER and then Base64-encoded before it's added to the transfer as a Memo.
+Note that this assumes that the ledger supports adding a memo to the transfer, but luckily, most ledgers do.
+
+## Payment tool
+
+The following is mainly a mix between the `pay.js` and `proxy.js` scripts from the Letter Shop tutorial,
+that can pay for content in reaction to an `Pay` header, and then stream that content to the console
+as it comes in. Save it as `tool.js`:
+```js
+const IlpPacket = require('ilp-packet')
+const Plugin = require('ilp-plugin-xrp-escrow')
+const crypto = require('crypto')
+const fetch = require('node-fetch')
+const uuid = require('uuid/v4')
+function base64 (buf) { return buf.toString('base64').replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, '') }
+function hash (secret) { return crypto.createHash('sha256').update(secret).digest() }
+function hmac (secret, input) { return crypto.createHmac('sha256', secret).update(input).digest() }
+
+const plugin = new Plugin({
+  secret: 'sndb5JDdyWiHZia9zv44zSr2itRy1',
+  account: 'rGtqDAJNTDMLaNNfq1RVYgPT8onFMj19Aj',
+  server: 'wss://s.altnet.rippletest.net:51233',
+  prefix: 'test.crypto.xrp.'
+})
+
+let counter = 0
+
+function sendTransfer (obj) {
+  obj.id = uuid()
+  obj.from = plugin.getAccount()
+  // to
+  obj.ledger = plugin.getInfo().prefix
+  // amount
+  // executionCondition
+  obj.expiresAt = new Date(new Date().getTime() + 1000000).toISOString()
+  return plugin.sendTransfer(obj).then(function () {
+    return obj.id
+  })
+}
+
+plugin.connect().then(function () {
+  return fetch('http://localhost:8000/')
+}).then(function (inRes) {
+  inRes.body.pipe(process.stdout)
+  const payHeaderParts = inRes.headers.get('Pay').split(' ')
+  console.log(payHeaderParts)
+  // e.g. Pay: 1 test.crypto.xrp.asdfaqefq3f.26wrgevaew SkTcFTZCBKgP6A6QOUVcwWCCgYIP4rJPHlIzreavHdU
+  setInterval(function () {
```

Comment:
Might not want that interval going forever...

Reply:
It goes on until you Ctrl-C it.